### PR TITLE
Register app as a way to share images

### PR DESCRIPTION
**Pull request by Sam1301**
This pr address #133 .
For now, I have added the functionality to share images, I will now work on sharing any type of files.
After sharing an image, the `ZulipActivity` starts and a `spinner` is shown during the upload. After uploading, the link of the image on server is added to the `messageEt` textView of `composeBox`.

**During upload**
screenshot_20161017-115042

**After upload is finished**
screenshot_20161017-115046

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@Sam1301, it looks like you've already signed the Dropbox CLA. Thanks!

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 1]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+        // create upload service client
+        Call<ResponseBody> call = ((ZulipApp) getApplicationContext()).getZulipServices().upload(description, body);
+        call.enqueue(new Callback<ResponseBody>() {
+            @Override
+            public void onResponse(Call<ResponseBody> call,
+                                   Response<ResponseBody> response) {
+                if (response.message().equals("OK")) {
+                    String filePathOnServer = "";
+                    try {
+                        String jsonResponse = getJsonResponseForUpload(response.body().charStream());
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: Thank you very much for the rigorous code view, I didn't know about gson and the best practices followed when consuming RESTful apis using retrofit. So I when through a couple of tutorials and made a 'UploadResponse` pojo and used it to deserialize the response.

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 2]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+            String loadingMsg = getResources().getString(R.string.uploading_message);
+            sendingMessage(true, loadingMsg);
+
+            // get actual file path
+            String imageFilePath = FilePathHelper.getPath(this, imageUri);
+
+            // upload the file asynchronously to the server
+            uploadFile(imageFilePath);
+        } else {
+            Toast.makeText(this, "Could not find the image", Toast.LENGTH_SHORT).show();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 3]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+            @Override
+            public void onResponse(Call<ResponseBody> call,
+                                   Response<ResponseBody> response) {
+                if (response.message().equals("OK")) {
+                    String filePathOnServer = "";
+                    try {
+                        String jsonResponse = getJsonResponseForUpload(response.body().charStream());
+                        filePathOnServer = parseUploadJSON(jsonResponse);
+                    } catch (IOException e) {
+                        e.printStackTrace();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 4]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+                    if (!filePathOnServer.equals("")) {
+                        // remove loading message from the screen
+                        sendingMessage(false, loadingMsg);
+
+                        // print message to compose box
+                        messageEt.append(" " + UrlHelper.addHost(filePathOnServer));
+                    } else {
+                        // remove loading message from the screen
+                        sendingMessage(false, loadingMsg);
+                        Toast.makeText(ZulipActivity.this, "Failed to upload", Toast.LENGTH_SHORT).show();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 5]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+                    }
+                }
+                else {
+                    // remove loading message from the screen
+                    sendingMessage(false, loadingMsg);
+                    Toast.makeText(ZulipActivity.this, "Failed to upload", Toast.LENGTH_SHORT).show();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 6]
```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
...
+                }
+
+            }
+
+            @Override
+            public void onFailure(Call<ResponseBody> call, Throwable t) {
+                // remove loading message from the screen
+                sendingMessage(false, loadingMsg);
+                Log.e("Upload error:", t.getMessage());
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 7]
```diff
@@ -879,15 +1023,19 @@ public void onTaskFailure(String result) {
...
         streamActv.setEnabled(!isSending);
         textView.setEnabled(!isSending);
         messageEt.setEnabled(!isSending);
         topicActv.setEnabled(!isSending);
         sendBtn.setEnabled(!isSending);
         togglePrivateStreamBtn.setEnabled(!isSending);
-        if (isSending)
+        if (isSending) {
+            Log.e("vvvvvvvvvvvvvvvvvv", "visible");
+            TextView msg = (TextView) composeStatus.getChildAt(1);
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: changed it :+1:

**Comment by Sam1301** [comment]
@niftynei Thanks so much for the rigorous code review, learnt a lot from it :)

**Comment by Sam1301** [comment]
For android 6.0 and higher, I have requested read extenal permission at run time.
screenshot_20161019-175125

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 8]
```diff
@@ -535,6 +540,142 @@ public Cursor runQuery(CharSequence charSequence) {
...
+                if (response.isSuccessful()) {
+                    String filePathOnServer = "";
+                        UploadResponse uploadResponse = response.body();
+                        filePathOnServer = uploadResponse.getUri();
+                    if (!filePathOnServer.equals("")) {
+                        // remove loading message from the screen
+                        sendingMessage(false, loadingMsg);
+
+                        // print message to compose box
+                        messageEt.append(" " + UrlHelper.addHost(filePathOnServer));
```
Comment (niftynei) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: This is a nit pick, but the urls that this is generating have two slashes in them at the seam between the host and the path: ie `https://zulip.tabbott.net//user_uploads/2/3c/lpFDn37AU7VWyJg4sZruImIX/IMG_20161015_155048.jpg`

**Comment by Sam1301** [comment]
I noticed that we have been using an extra forward slash in `RecyclerMessageAdapter.setUpGravatar()` too. The problem traces back to `UrlHelper.addHost()` which should return host url without an appended '/' in the end. Since this change was in the existing codebase, I made a different commit for it :)

**Comment by niftynei** [comment]
Hey @Sam1301 thanks for the fixes. This is looking really good. I found one large issue that I'm not sure is a blocker, but I'd like to get your opinion on before merging this in.

Basically the problem is that if the app is open and in the background but the activity hasn't been destroyed yet, sharing to Zulip fails.  The reason for this is that the code to kick off a share intent is in `onCreate`, since the activity was never destroyed (just Stopped), then it's never created again.

I added some logging to onCreate, onStart & onResume so you can see this happen in the logs:

Here's a share where the activity has been destroyed:
```
10-19 15:25:49.026 11146 11146 I status  : suspend
10-19 15:25:55.162 11146 11146 I status  : create
10-19 15:25:55.441 11146 11146 I status  : onStart
10-19 15:25:55.441 11146 11146 I status  : Intent { act=android.intent.action.SEND typ=image/jpeg flg=0x10000001 cmp=com.zulip.android.dev/com.zulip.android.activities.ZulipActivity clip={image/jpeg U:content://com.google.android.apps.photos.contentprovider/0/1/content%3A%2F%2Fmedia%2Fexternal%2Fimages%2Fmedia%2F204876/REQUIRE_ORIGINAL/NONE/551220734} (has extras) }
10-19 15:25:55.446 11146 11146 I status  : resume
10-19 15:25:55.446 11146 11146 I status  : android.intent.action.SEND
```

And here's a share where the activity has been backgrounded, but not completely destroyed yet.  Note that create wasn't called.  In this case, the logic to upload the shared file isn't triggered.
```
10-19 15:27:01.560 11146 11146 I status  : suspend
10-19 15:27:17.736 11146 11146 I status  : onStart
10-19 15:27:17.737 11146 11146 I status  : Intent { act=android.intent.action.SEND typ=image/jpeg flg=0x10000001 cmp=com.zulip.android.dev/com.zulip.android.activities.ZulipActivity clip={image/jpeg U:content://com.google.android.apps.photos.contentprovider/0/1/content%3A%2F%2Fmedia%2Fexternal%2Fimages%2Fmedia%2F204876/REQUIRE_ORIGINAL/NONE/551220734} (has extras) }
10-19 15:27:17.743 11146 11146 I status  : resume
10-19 15:27:17.743 11146 11146 I status  : android.intent.action.SEND
```

I think there's a few ways to fix this, the best suggestion being to use `onNewIntent`, which gets called when a new Intent is sent to the Activity.  Here's a log of onNewIntent being called even when onCreate isn't:

```
10-19 15:43:52.519 14476 14476 I status  : suspend
10-19 15:44:05.646 14476 14476 I status  : onNewIntent
10-19 15:44:05.728 14476 14476 I status  : onStart
10-19 15:44:05.729 14476 14476 I status  : Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.zulip.android.dev/com.zulip.android.activities.ZulipActivity bnds=[13,811][224,1053] (has extras) }
10-19 15:44:05.738 14476 14476 I status  : onResume
```

**Comment by Sam1301** [comment]
Thanks, I didn't think about this test case, will make necessary changes to the pr :+1:

**Comment by Sam1301** [comment]
Made the suggested changes, now the user can upload images even when the activity hasn't been destroyed.

**Comment by niftynei** [comment]
Thanks! This has been merged.

**Comment by Sam1301** [comment]
Should I try to extend this functionality to all types of files or will it be better to share all types of files using attachment uploads?

**Comment by niftynei** [comment]
Extending it to all file types sounds good to me, but there's a lot to this.  The one thing I think it's really missing right now is being able to share text from other apps as well.

**Comment by Sam1301** [comment]
I agree, I would love to give it a try 👍

**Pull request closed without merging**